nixos/lib/test-driver/test-driver.py

```diff
@@ -525,6 +526,36 @@ def screenshot(self, filename: str) -> None:
             if ret.returncode != 0:
                 raise Exception("Cannot convert screenshot")
 
+    def copy_from_vm(self, source: str, target_dir: str = "") -> None:
+        """Copy a file from the VM (specified by an in-VM source path) to $out
+        (with an optional subdirectory parameter). The file is copied via the
+        shared directory.
+        """
+        # Compute the source, target, and intermediate shared file names
+        out_dir = os.environ.get("out", os.getcwd())
+        filename = re.sub(r".*/", "", source)
+        int_shared_dir = "/tmp/xchg"
+        shared_temp = tempfile.mkdtemp(dir=self.shared_dir)
+        shared_temp_basename = re.sub(r".*/", "", shared_temp)
+        shared_temp_in = "{}/{}".format(int_shared_dir, shared_temp_basename)
+        src_out = "{}/{}".format(shared_temp, filename)
+        src_in = "{}/{}".format(shared_temp_in, filename)
+        tgt = "{}/{}/".format(out_dir, target_dir)
+        # Copy the file to the shared directory inside VM
+        if not self.is_up():
```

Comment:
current behaviour in the `Machine` class is to start machines if you request things from them, so we should do here, too.